src/main/java/teammates/common/datatransfer/attributes/AccountAttributes.java

```diff
@@ -26,49 +27,109 @@
 
-    public AccountAttributes(String googleId, String name, boolean isInstructor,
-                String email, String institute, StudentProfileAttributes studentProfileAttributes) {
+    public AccountAttributes(String googleId, String name, boolean isInstructor, String email, String institute) {
         this.googleId = SanitizationHelper.sanitizeGoogleId(googleId);
         this.name = SanitizationHelper.sanitizeName(name);
         this.isInstructor = isInstructor;
         this.email = SanitizationHelper.sanitizeEmail(email);
         this.institute = SanitizationHelper.sanitizeTitle(institute);
-        this.studentProfile = studentProfileAttributes;
-        this.studentProfile.sanitizeForSaving();
+    }
 
+    public static Builder builder(String googleId, String name, boolean isInstructor, String email, String institute) {
+        return new Builder(googleId, name, isInstructor, email, institute);
     }
 
-    public AccountAttributes(String googleId, String name, boolean isInstructor,
-                String email, String institute) {
-        this.googleId = SanitizationHelper.sanitizeGoogleId(googleId);
-        this.name = SanitizationHelper.sanitizeName(name);
-        this.isInstructor = isInstructor;
-        this.email = SanitizationHelper.sanitizeEmail(email);
-        this.institute = SanitizationHelper.sanitizeTitle(institute);
-        this.studentProfile = StudentProfileAttributes.builder().build();
-        this.studentProfile.googleId = this.googleId;
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        private static final String REQUIRED_FIELD_CANNOT_BE_NULL = "Non-null value expected";
+        private AccountAttributes accountAttributes;
+
+        public Builder(String googleId, String name, boolean isInstructor, String email, String institute) {
+            validateRequiredFields(googleId, name, isInstructor, email, institute);
+            accountAttributes = new AccountAttributes(googleId, name, isInstructor, email, institute);
+        }
+
+        public Builder() {
+            accountAttributes = new AccountAttributes();
+        }
+
+        public AccountAttributes build() {
+            return accountAttributes;
+        }
+
+        public Builder withCreatedAt(Date createdAt) {
+            if (createdAt != null) {
+                accountAttributes.createdAt = createdAt;
+            }
+            return this;
+        }
+
+        public Builder withStudentProfile(StudentProfile studentProfile) {
+            accountAttributes.studentProfile =
+                    studentProfile == null ? null : StudentProfileAttributes.valueOf(studentProfile);
+            return this;
+        }
+
+        public Builder withStudentProfileAttributes(StudentProfileAttributes studentProfileAttributes) {
+            accountAttributes.studentProfile = studentProfileAttributes;
+            accountAttributes.studentProfile.sanitizeForSaving();
```

Comment:
While initializing data members, some times the parameter is sanitized and sometimes not. So I will need to add two `withXXX()` methods:

- `withGoogleIdSanitized()`
- `withGoogleId()`

Reply:
I think that might be a tech-debt for inconsistency in sanitization rather than an intended behaviour. Can you confirm this?

Reply:
1. For AccountAttributes, two constructors were used, one with no parameter so all the fields were assigned without sanitizing. Another one was with 5 parameters, where 4 fields were sanitized( all strings). Can I confirm that you currently think that this inconsistent and all might be need to be sanitzed at all times( hence pushing it to `build()` and not requiring two different `wiithXXX()` for sanitized and not sanitized)?

1.  Regarding Sanitization inside the StudentProfileAtributes, I have put it inside `build()` as it had the same behaviour in both `withSPA()`

Reply:
1. You already have done that! Good job!
2. My suggestion: We can write a null check in the `build()` method to see whether we should `sanitizeForSaving` for `studentProfile`. In doing this, it will be more consistent as we do all the sanitizations in the last `build` method.